### DISCUSSION TITLE
Addition of ECG as SleepStaging prediction input data

### DISCUSSION THREAD

**Discussion by LaurentLM** [raphaelvallat/yasa#183] [Ideas]
Hi Raphaël,

I have been playing with most of the YASA features, but with a particular focus on sleep stage prediction, hypnograms and hrv_stage. I must admit, I am very impressed by the range of measurements it offers and the quality of the documentation. A big thank you for making this public and so easy to use.

yasa.SleepStaging currently makes use of EEG, EMG and EOG to predict sleep stages. I was wondering if adding ECG as an optional signal for yasa.SleepStaging would be likely to bring any improvements to the model performance? 
[SleepECG](https://github.com/cbrnr/sleepecg) showed they can determine sleep stages based on ECG alone. I have experimented by feeding ECG as EMG data (based on the fact EMGs tend to be strongly impacted by ECG) and saw some improvements in the model probabilities.

Is that be a feature you would look into integrating in the future? 
Disclaimer, I not familiar with building ML models at all and have no idea how much this represents. It seems sleep staged datasets with ECG are available, which has to be a good start :-)

**Comment by raphaelvallat** [comment]
Hi @LaurentLM,

Thank you for the kind words, and glad you have found YASA to be helpful for your own work! This is a great question. In my own work I have experimented with ML model to predict sleep stages from ECG alone, and can confirm that these models can indeed reach high-level of accuracy. However, I don't really think they would bring any significant improvements compared to using EEG, EOG and EMG. My guess is that at best they will add 1-2% accuracy. 

Now, the main issue is that it would require a significant effort to add processing of the ECG in YASA's training datasets (see [yasa_classifier](https://github.com/raphaelvallat/yasa_classifier)), and derive relevant features from the ECG to be used in the ML model (beyond just the heart rate). I don't have this kind of time now unfortunately. That said, I have been hoping to do a major upgrade of the YASA SleepStaging model in the coming months, so maybe I could try to squeeze that in at that time.

Note that for best slepe staging accuracy, I would recommend using the [U-sleep model](https://sleep.ai.ku.dk/sleep_stager#), and/or a consensus of U-sleep and YASA.
